MH template and sites that use it

Robonau opened the pull request:
resolves #2088
resolves #2469
also changed Kuimh to use template

Robonau committed:
MH template and caisemh

resolves #2088
also changed Kuimh to use template

Robonau committed:
Create caisemh

icon

27136773 commented on src/web/mjs/connectors/templates/MH.mjs:
```diff
@@ -0,0 +1,69 @@
+        this.url = undefined;
+
+        this.path = '/booklist?page=';
+        this.pathMatch = 'page=(\\d+)';
```
I'm surprised this works since this is a string.
However i would still recommend to use a regex literal.
`/page=(\d+)/`

27136773 replied:
May also rename to `queryMangasPageCountMatcher`, to imply that this belongs to the logic that extract the manga page count from the pagination link

Robonau replied:
that's what Kuimh was before, il change anyways

27136773 commented on src/web/mjs/connectors/templates/MH.mjs:
```diff
@@ -0,0 +1,69 @@
+        this.pathMatch = 'page=(\\d+)';
+        this.queryMangasPageCount = 'div.page-pagination div.pagination li:nth-last-child(2) a';
+        this.queryMangas = 'ul.mh-list li div.mh-item-detali h2.title a';
+        this.queryMangasURL = 'section.banner_detail div.info h1';
```
Member name implies to be used in `_getMangas`, but in fact this is used to extract a title...
May rename to something like `queryMangaTitle`

27136773 commented on src/web/mjs/connectors/Caisemh.mjs:
```diff
@@ -0,0 +1,23 @@
+    async _getPages(chapter) {
+        this.requestOptions.headers.set('x-user-agent', 'Mozilla/5.0 (Linux; Android 9; Pixel) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4026.0 Mobile Safari/537.36');
+        let request = new Request(new URL(chapter.id, this.url), this.requestOptions);
+        let data = await this.fetchDOM(request, this.queryPages);
```
Should not await, otherwise any other request made by this connector will use the mobile user agent that is set until the await is completed.

Either use a promise to continue instantly, or even better don't modify the user agent in the global request options and instead declare some new request options, e.g.
```javascript
const request = new Request(new URL(chapter.id, this.url), {
    headers: {
        'x-user-agent': 'Mozilla/5.0 Android X'
    }
});
```

Robonau committed:
fixes

/page=(\d+)/
queryMangaTitle
request options

Robonau committed:
added mkzhan

ronny mentioned on discord
added to this pull cause the template hasn't gone through yet

Robonau committed:
icon for mkzhan

Robonau committed:
lint fixes

Robonau committed:
iqiyi

http://www.iqiyi.com/manhua
mentioned on discord by ronny
uses MH template

Robonau committed:
Update iqiyi.mjs

consolelog

Robonau committed:
Update Kuman5.mjs

kuman55 using HM
not that it keeps anything but _getMangaFromURI from it

27136773 commented on src/web/mjs/connectors/Caisemh.mjs:
```diff
@@ -0,0 +1,24 @@
+    async _getPages(chapter) {
+        const request = new Request(new URL(chapter.id, this.url), {
+            headers: {
+                'x-user-agent': 'Mozilla/5.0 Android X'
```
Does this work with this user agent?
I was just suggesting with some random string.

Robonau replied:
it works fine, but il change it back to the useragent that i had before

27136773 replied:
It is possible that the server just checked if the user agent contains `ios|android`, so it probably worked with this minimalistic user agent

27136773 commented on src/web/mjs/connectors/templates/MH.mjs:
```diff
@@ -0,0 +1,69 @@
+        this.tags = [];
+        this.url = undefined;
+
+        this.path = '/booklist?page=';
```
May use a placeholder to allow better control for websites that do not use the page
number as last element, e.g.
`'/booklist?page=%PAGE%';`

27136773 commented on src/web/mjs/connectors/templates/MH.mjs:
```diff
@@ -0,0 +1,69 @@
+
+    async _getMangas() {
+        let mangaList = [];
+        let request = new Request(new URL(this.path, this.url), this.requestOptions);
```
In case of the placeholder solution change this to e.g.
`new URL(this.path.replace('%PAGE%', 1), this.url)`

27136773 commented on src/web/mjs/connectors/templates/MH.mjs:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    async _getMangasFromPage(page) {
+        let request = new Request(new URL(this.path + page, this.url), this.requestOptions);
```
In case of the placeholder solution change this to e.g.
`new URL(this.path.replace('%PAGE%', page), this.url)`

27136773 commented on src/web/mjs/connectors/Kuman5.mjs:
```diff
@@ -1,21 +1,14 @@
         super.label = '酷漫屋 (Kuman55)';
         this.tags = [ 'manga', 'chinese' ];
         this.url = 'http://www.kuman55.com';
-    }
```
In case of the placeholder solution, overwrite the path, e.g.
`this.path = '/sort/1-%PAGE%.html';`

27136773 commented on src/web/mjs/connectors/Kuman5.mjs:
```diff
@@ -29,7 +22,7 @@ export default class Kuman5 extends Connector {
 
     async _getMangasFromPage(page) {
```
In case of the placeholder solution this method no longer needs to be overwritten, the base from template should work

27136773 commented on src/web/mjs/connectors/Kuman5.mjs:
```diff
@@ -1,21 +1,14 @@
+        this.queryChapter = 'div#chapterlistload ul#detail-list-select-1 li a';
     }
 
     async _getMangas() {
```
Apply the customized chapter IDs before returning the list, e.g.
```javascript
mangaList.forEach(manga => manga.id = `/mulu${manga.id}1-1.html`);
return mangaList;
When this is done, there is n o longer a need to overwrite the `_getChapters` method from the base class.
```

27136773 commented on src/web/mjs/connectors/Kuman5.mjs:
```diff
@@ -40,7 +33,7 @@ export default class Kuman5 extends Connector {
 
     async _getChapters(manga) {
```
In case of applying the customized chapter IDs in `_getMangas` method, this method no longer needs to be overwritten, the base from template should work

Robonau committed:
fixes

Robonau committed:
Create XianMan123.mjs

there is probably a way to have that regex be just one, but i aint too good with regex
reoslves #2469

Robonau committed:
Update Kuman5.mjs

Robonau committed:
Update XianMan123.mjs

console.log

Robonau committed:
Update XianMan123.mjs

Robonau committed:
Create xianman123

icon

Robonau committed:
Merge remote-tracking branch 'upstream/master' into caisemh

27136773 commented on src/web/mjs/connectors/templates/MH.mjs:
```diff
@@ -0,0 +1,69 @@
+    async _getPages(chapter) {
+        let request = new Request(new URL(chapter.id, this.url), this.requestOptions);
+        let data = await this.fetchDOM(request, this.queryPages);
+        return data.map(element => this.getAbsolutePath(element.dataset.echo || element, request.url));
```
May extend options to cover cases from `iqiyi ` and `mkzhan`, so these websites don't need to overwrite it
`this.getAbsolutePath(element.dataset.src || element.dataset.original || element.dataset.echo || element, request.url)`

27136773 commented on src/web/mjs/connectors/mkzhan.mjs:
```diff
@@ -0,0 +1,36 @@
+        this.queryPages = 'div.rd-article-wr div.rd-article__pic source';
+    }
+
+    async _getPages(chapter) {
```
May merge logic into template and remove this method

27136773 commented on src/web/mjs/connectors/iqiyi.mjs:
```diff
@@ -0,0 +1,49 @@
+
+    }
+
+    async _getPages(chapter) {
```
May merge logic into template and remove this method

27136773 commented on src/web/mjs/connectors/mkzhan.mjs:
```diff
@@ -0,0 +1,36 @@
+        super.id = 'mkzhan';
+        super.label = 'mkzhan';
+        this.tags = [ 'manga', 'webtoon', 'chinese' ];
+        this.url = 'https://www.mkzhan.com/';
```
Remove trailing `/`

Robonau committed:
fixes

27136773 commented on src/web/mjs/connectors/iqiyi.mjs:
```diff
@@ -0,0 +1,49 @@
+        super.id = 'iqiyi';
+        super.label = 'iqiyi';
+        this.tags = [ 'manga', 'webtoon', 'chinese' ];
+        this.url = 'https://www.iqiyi.com/';
```
Remove trailing `/`

Robonau committed:
Update iqiyi.mjs

tailing /

27136773 reviewed (commented):
Ready for merge, as soon as the Lint errors have been fixed

Robonau committed:
lint

27136773 reviewed (approved):
Looks good
